examples/metrics/application_efficiency.py

```diff
@@ -0,0 +1,301 @@
+#
+# For example, plotting a bar chart of application efficiences for one
+# application can help us to summarize that application's performance more
+# effectively than a table:
+
+filtered = effs[effs["application"]=="MyApp"]
+filtered.plot(kind="bar", x="platform", y="app eff", xlabel="Platform", ylabel="Application Efficiency", legend=False)
+plt.savefig("application_efficiency_bars_2023.png")
+
+# %%
+# We can now clearly see that MyApp can adapt to and make very effective use of
+# Platforms A and B, is within 2x of state-of-the-art performance on Platform
+# C, but performs poorly on Platforms D and E. The key takeaway from this
+# analysis is that a developer wishing to improve the "performance portability"
+# of MyApp should focus on improving support for Platforms D and E.
+#
+# Working with Historical Data
+# ----------------------------
+#
+# The performance of an application can change over time, as developers add new
+# features and optimize for new platforms, or due to changes in the software
+# stack (e.g., new compiler or driver versions).
+#
+# Let's see how this could affect the application efficiency of MyApp, by adding
+# some new data points collected at a later point in time:
+#
+# .. list-table::
+#     :widths: 20 20 20 20 20
+#     :header-rows: 1
+#
+#     * - problem
+#       - application
+#       - platform
+#       - fom
+#       - date
+#
+#     * - Test
+#       - MyApp
+#       - A
+#       - 30.0
+#       - 2024
+#     * - Test
+#       - MyApp
+#       - B
+#       - 15.0
+#       - 2024
+#     * - Test
+#       - MyApp
+#       - C
+#       - 25.0
+#       - 2024
+#     * - Test
+#       - MyApp
+#       - D
+#       - 3.0
+#       - 2024
+#     * - Test
+#       - MyApp
+#       - E
+#       - 2.5
+#       - 2024
+
+# %%
+# We can compute application efficiency as before:
+
+#sphinx_gallery_start_ignore
+new_myapp_data = {
+    "problem": ["Test"] * 5,
+    "application": ["MyApp"] * 5,
+    "platform": ["A", "B", "C", "D", "E"],
+    "fom": [30.0, 15.0, 25.0, 3.0, 2.5],
+    "date": [2024] * 5,
+}
+
+df = pd.concat([df, pd.DataFrame(new_myapp_data)], ignore_index=True)
+#sphinx_gallery_end_ignore
+
+effs = p3.metrics.application_efficiency(df)
+print(effs)
+
+# %%
+# These latest results suggest that the developers of MyApp acted upon
+# earlier results and improved support for Platforms D and E. But in doing so,
+# a small performance regression was introduced in Platforms A and B.
+#
+# .. note::
+#     Such trade-offs are very common, especially when developers wish to
+#     maintain a single source code that targets multiple platforms.
+#
+# Computing the correct application efficiency values for MyApp and YourApp
+# requires that our dataset contains all of our historical performance results.
+# Since what we're really interested in understanding is the *latest*
+# application efficiency, we should take care to filter our data appropriately
+# before producing any plots.
```

Comment:
Do you also want to finish the tutorial on a note about how `date` can be used with some other tracking thing like a hash?

Reply:
I've added a conclusion of sorts in 5c1d53e895f7b4f3a75633b638c98401c802e03f.  I also had to change the sphinx configuration to make this work (see afe550a6cfa9c140931e7eb1372055d33af7dd41).  Let me know what you think.